### PR TITLE
Improve CMake

### PR DESCRIPTION
**Pull request by Warchant**
- make build of executables optional (enabled by default, but can be disabled)
- make static libraries at installation having `binaryen-` preffix (to prevent name collisions, so names will be `libbinaryen-ir`, etc)
- export cmake config, so other projects can easily import binaryen (as `binaryen::passes`, or `binaryen::binaryen` imported targets)
- install all headers from src to `include/binaryen` dir

This version is going to be [hunterized](https://docs.hunter.sh/en/latest/creating-new/create/cmake.html). 

UPD: https://docs.hunter.sh/en/latest/packages/pkg/binaryen.html

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -167,126 +168,155 @@ IF(BUILD_STATIC_LIB)
 ELSE()
   ADD_LIBRARY(binaryen SHARED ${binaryen_SOURCES})
 ENDIF()
-TARGET_LINK_LIBRARIES(binaryen passes wasm asmjs emscripten-optimizer ir cfg support)
-INSTALL(TARGETS binaryen DESTINATION ${CMAKE_INSTALL_LIBDIR})
 
-INSTALL(FILES src/binaryen-c.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+set(_dependencies "passes;wasm;asmjs;emscripten-optimizer;ir;cfg;support")
+foreach(_dep ${_dependencies})
```
Comment (sbc100) [inline CMakeLists.txt]: the convention so far in this file seems to be ALL_CAPS for cmake stuff.  Probably better to follow the local convention for changes like this.

File: CMakeLists.txt [hunk 2]
```diff
@@ -167,126 +168,155 @@ IF(BUILD_STATIC_LIB)
...
+
+TARGET_LINK_LIBRARIES(binaryen ${_dependencies})
+install(TARGETS binaryen
+  EXPORT binaryenConfig
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+  INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+  PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+  FRAMEWORK DESTINATION ${CMAKE_INSTALL_PREFIX}
```
Comment (sbc100) [inline CMakeLists.txt]: Seems like this whole block is duplicated in a lot of places.. and that these values should really be the defaults.  Are they not?

**Comment by Warchant** [comment]
>I'm not convinced we really need to be installing static libraries and binaryen headers other than the one we already install ("binaryen-c.h")

My colleagues assure me that they need to inherit from `ShellExternalInterface` from `shell-interface.h` to execute C++ function from wasm:
https://github.com/soramitsu/polkadot/blob/master/test/deps/binaryen/binaryen_test.cpp#L18

Do you have any code example, how to mimic that logic with `binaryen-c.h`

**Comment by sbc100** [comment]
> > I'm not convinced we really need to be installing static libraries and binaryen headers other than the one we already install ("binaryen-c.h")
> 
> My colleagues assure me that they need to inherit from `ShellExternalInterface` from `shell-interface.h` to execute C++ function from wasm:
> https://github.com/soramitsu/polkadot/blob/master/test/deps/binaryen/binaryen_test.cpp#L18
> 

I see. Do they also have the requirement that they need to work from a "packaged" or "installed" version of binaryen, or can they use it directly from a source archive or git checkout?

**Comment by Warchant** [comment]
>I see. Do they also have the requirement that they need to work from a "packaged" or "installed" version of binaryen, or can they use it directly from a source archive or git checkout?

We prefer [hunter](https://hunter.sh/) for dependency management :) 

It is the same as 
```
wget https://github.com/WebAssembly/binaryen/archive/1.38.28.tar.gz
tar -xf 1.38.28.tar.gz
cd 1.38.28
mkdir build
cd build
cmake .. 
make install
```

**Comment by Warchant** [comment]
@sbc100 any thoughts on the problem I've described above?

**Comment by sbc100** [comment]
Can you respond to my earlier comments, and also split the PR as I suggested?   Then we can revisit.

**Pull request closed without merging**